guardian/pkg/server/server.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	calicotls "github.com/projectcalico/calico/crypto/pkg/tls"
+	"github.com/projectcalico/calico/guardian/pkg/conn"
+	"github.com/projectcalico/calico/guardian/pkg/tunnel"
+)
+
+type Server interface {
+	ListenAndServeCluster(ctx context.Context) error
+	ListenAndServeManagementCluster(ctx context.Context) error
+}
+
+// Client is the voltron client. It is used by Guardian to establish a secure tunnel connection to the Voltron server and
+// then enable managed cluster services and management cluster services to communicate with one another.
+type server struct {
+	http     *http.Server
+	proxyMux *http.ServeMux
+	targets  []Target
+
+	tunnelAddr string
+	tunnelCert *tls.Certificate
+
+	// tunnelRootCAs defines the set of root certificate authorities that guardian will use when verifying voltron certificates.
+	// if nil, dialer will use the host's CA set.
+	tunnelRootCAs *x509.CertPool
+	// TunnelServerName defines the server name to be used when connecting to Voltron
+	tunnelServerName string
+
+	tunnel tunnel.Tunnel
+
+	connRetryAttempts int
+	connRetryInterval time.Duration
+
+	listenPort string
+	listenHost string
+
+	tunnelDialerOptions []tunnel.DialerOption
+}
+
+func New(addr string, opts ...Option) (Server, error) {
+	var err error
+	srv := &server{
+		http: new(http.Server),
+
+		tunnelAddr:        addr,
+		tunnelServerName:  strings.Split(addr, ":")[0],
```

Comment:
What if we have an IPV6 address ?

Reply:
This isn't splitting an IP, the address is a domain name.

Reply:
I've taken this out, it's not actually needed here.